packages/datadog-plugin-fs/src/index.js

```diff
@@ -0,0 +1,490 @@
+'use strict'
+
+let kDirReadPromisified
+let kDirClosePromisified
+
+const tagMakers = {
+  open: createOpenTags,
+  close: createCloseTags,
+  readFile: createReadFileTags,
+  writeFile: createWriteFileTags,
+  appendFile: createAppendFileTags,
+  access: createPathTags,
+  copyFile: createCopyFileTags,
+  stat: createPathTags,
+  lstat: createPathTags,
+  fstat: createFDTags,
+  readdir: createPathTags,
+  opendir: createPathTags,
+  read: createFDTags,
+  write: createFDTags,
+  writev: createFDTags,
+  chmod: createChmodTags,
+  lchmod: createChmodTags,
+  fchmod: createFchmodTags,
+  chown: createChownTags,
+  lchown: createChownTags,
+  fchown: createFchownTags,
+  realpath: createPathTags,
+  readlink: createPathTags,
+  unlink: createPathTags,
+  symlink: createCopyFileTags,
+  link: createCopyFileTags,
+  rmdir: createPathTags,
+  rename: createCopyFileTags,
+  fsync: createFDTags,
+  fdatasync: createFDTags,
+  mkdir: createPathTags,
+  truncate: createPathTags,
+  ftruncate: createFDTags,
+  utimes: createPathTags,
+  futimes: createFDTags,
+  mkdtemp: createPathTags
+}
+
+function createWrapCreateReadStream (config, tracer) {
+  return function wrapCreateReadStream (createReadStream) {
+    return function wrappedCreateReadStream (path, options) {
+      const tags = makeFSFlagTags(path, options, 'r', config, tracer)
+      return tracer.trace('fs.readstream', { tags }, (span, done) => {
+        const stream = createReadStream.apply(this, arguments)
+        stream.once('end', done)
+        stream.once('error', done)
+        return stream
+      })
+    }
+  }
+}
+
+function createWrapCreateWriteStream (config, tracer) {
+  return function wrapCreateWriteStream (createWriteStream) {
+    return function wrappedCreateWriteStream (path, options) {
+      const tags = makeFSFlagTags(path, options, 'w', config, tracer)
+      return tracer.trace('fs.writestream', { tags }, (span, done) => {
+        const stream = createWriteStream.apply(this, arguments)
+        stream.once('finish', done)
+        stream.once('error', done)
+        return stream
+      })
+    }
+  }
+}
+
+function createWrapExists (config, tracer) {
+  return function wrapExists (exists) {
+    return function wrappedExists (path, cb) {
+      if (typeof cb !== 'function') {
+        return exists.apply(this, arguments)
+      }
+      const tags = makeFSTags(path, null, config, tracer)
+      return tracer.trace('fs.exists', { tags }, (span, done) => {
+        arguments[1] = function (result) {
+          done()
+          cb.apply(this, arguments)
+        }
+        return exists.apply(this, arguments)
+      })
+    }
+  }
+}
+
+function createWrapDirRead (config, tracer, sync) {
+  const name = sync ? 'fs.dir.readsync' : 'fs.dir.read'
+  return function wrapDirRead (read) {
+    function options () {
+      const tags = makeFSTags(this.path, null, config, tracer)
+      return { tags }
+    }
+    return tracer.wrap(name, options, read)
```

Comment:
Should we wrap only when there is a parent? Otherwise requires and any files read/written outside of a request will be traces which may be unexpected and/or unwanted. (applies to all calls to `tracer.wrap`).